### ISSUE TITLE
Crash for some users Realm 4.1.1 - realm::util::terminate(char const*, char const*, long) + 10

### ISSUE THREAD

**Issue by Apach3Q** [realm/realm-cocoa#6863]
## Goals
I want update realm database, include insert or update or delete operations.
This problem has been bothering me for a long time, but I can't solve it.

## Expected Results
No crash.

## Actual Results
```
Crashed: com.apple.main-thread
0  libsystem_kernel.dylib         0x19763dd88 __pthread_kill + 8
1  libsystem_pthread.dylib        0x1975561e8 pthread_kill$VARIANT$mp + 136
2  libsystem_c.dylib              0x1974a9644 abort + 100
3  Realm                          0x104204a1c realm::util::terminate(char const*, char const*, long) + 10
4  Realm                          0x104204cdc realm::util::terminate_internal(std::__1::basic_stringstream<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) + 704
5  Realm                          0x104204e54 realm::util::terminate(char const*, char const*, long, std::initializer_list<realm::util::Printable>&&) + 328
6  Realm                          0x103f3cfd4 realm::_impl::RealmCoordinator::RealmCoordinator() + 182
7  Realm                          0x104012960 realm::Realm::cache_new_schema() + 100
8  Realm                          0x10401325c realm::Realm::begin_transaction() + 176
9  Realm                          0x103fe1acc -[RLMRealm beginWriteTransaction] + 7504
10 MeviiBibleRead                 0x10249ef98 specialized RLMRealm.commitTransaction(block:) + 4366806936
11 MeviiBibleRead                 0x10249fc84 static RLMRealm.insertOrUpdateItem(_:) + 4366810244
12 MeviiBibleRead                 0x10268a33c TimeRecordManager.createTodayNewRecordModel(stayTime:readingTime:vodTime:dodTime:bibleTime:listeningTime:quizTime:) + 285 (TimeRecordManager.swift:285)
13 MeviiBibleRead                 0x102689b10 TimeRecordManager.endReading() + 196 (TimeRecordManager.swift:196)
14 MeviiBibleRead                 0x10250f538 closure #2 in HomeViewController.bindModel() + 225 (HomeViewController.swift:225)
15 MeviiBibleRead                 0x102510660 partial apply for thunk for @escaping @callee_guaranteed (@unowned HomeTopbarType) -> () + 4367271520 (<compiler-generated>:4367271520)
16 RxSwift                        0x104619b54 closure #1 in ObservableType.subscribe(onNext:onError:onCompleted:onDisposed:) + 65 (ObservableType+Extensions.swift:65)
17 RxSwift                        0x104619d94 partial apply for closure #1 in ObservableType.subscribe(onNext:onError:onCompleted:onDisposed:) + 4401831316 (<compiler-generated>:4401831316)
18 RxSwift                        0x1045cada0 AnonymousObserver.onCore(_:) + 22 (AnonymousObserver.swift:22)
19 RxSwift                        0x10461c7ac ObserverBase.on(_:) + 16 (ObserverBase.swift:16)
20 RxSwift                        0x10461c910 protocol witness for ObserverType.on(_:) in conformance ObserverBase<A> + 4401842448 (<compiler-generated>:4401842448)
21 RxSwift                        0x10463d09c Sink.forwardOn(_:) + 35 (Sink.swift:35)
22 RxSwift                        0x104653844 WithLatestFromSink._synchronized_on(_:) + 79 (WithLatestFrom.swift:79)
23 RxSwift                        0x104643c1c SynchronizedOnType.synchronizedOn(_:) + 15 (<compiler-generated>:15)
24 RxSwift                        0x1046534c4 WithLatestFromSink.on(_:) + 69 (WithLatestFrom.swift:69)
25 RxSwift                        0x104653a28 protocol witness for ObserverType.on(_:) in conformance WithLatestFromSink<A, B, C> + 4402068008 (<compiler-generated>:4402068008)
26 RxSwift                        0x10463d09c Sink.forwardOn(_:) + 35 (Sink.swift:35)
27 RxSwift                        0x1046405d8 SubscribeOnSink.on(_:) + 46 (SubscribeOn.swift:46)
28 RxSwift                        0x104640c4c protocol witness for ObserverType.on(_:) in conformance SubscribeOnSink<A, B> + 4401990732 (<compiler-generated>:4401990732)
29 RxSwift                        0x10463d09c Sink.forwardOn(_:) + 35 (Sink.swift:35)
30 RxSwift                        0x10460f674 MapSink.on(_:) + 45 (Map.swift:45)
31 RxSwift                        0x10460f7d8 protocol witness for ObserverType.on(_:) in conformance MapSink<A, B> + 4401788888 (<compiler-generated>:4401788888)
32 RxSwift                        0x10461cd20 partial apply + 4401843488 (<compiler-generated>:4401843488)
33 RxSwift                        0x104637234 thunk for @escaping @callee_guaranteed (@in_guaranteed Event<A>) -> () + 4401951284 (<compiler-generated>:4401951284)
34 RxSwift                        0x104622ffc PublishSubject.on(_:) + 14 (Bag+Rx.swift:14)
35 RxCocoa                        0x1044de860 @objc MessageSentProxy.methodInvoked(withArguments:) + 194072 (<compiler-generated>:194072)
36 RxCocoa                        0x1044af760 RX_forward_invocation + 234 (_RXObjCRuntime.m:234)
37 RxCocoa                        0x1044af60c __71-[RXObjCRuntime(InfrastructureMethods) swizzleForwardInvocation:error:]_block_invoke_2 + 482 (_RXObjCRuntime.m:482)
38 CoreFoundation                 0x19784c414 ___forwarding___ + 676
39 CoreFoundation                 0x19784e5bc _CF_forwarding_prep_0 + 92
40 UIKitCore                      0x19b2b7fc0 -[UIViewController _setViewAppearState:isAnimating:] + 832
41 UIKitCore                      0x19b21b34c -[UINavigationController viewDidAppear:] + 188
42 UIKitCore                      0x19b2b7fc0 -[UIViewController _setViewAppearState:isAnimating:] + 832
43 UIKitCore                      0x19b1f2734 -[UITabBarController viewDidAppear:] + 124
44 UIKitCore                      0x19b2b7fc0 -[UIViewController _setViewAppearState:isAnimating:] + 832
45 UIKitCore                      0x19b2b89dc -[UIViewController _endAppearanceTransition:] + 228
46 UIKitCore                      0x19b1d25c4 __48-[UIPresentationController transitionDidFinish:]_block_invoke + 136
47 UIKitCore                      0x19b8fc94c _runAfterCACommitDeferredBlocks + 316
48 UIKitCore                      0x19b8ec2f4 _cleanUpAfterCAFlushAndRunDeferredBlocks + 232
49 UIKitCore                      0x19b91c744 _afterCACommitHandler + 76
50 CoreFoundation                 0x1977c5fb8 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 32
51 CoreFoundation                 0x1977c0eac __CFRunLoopDoObservers + 420
52 CoreFoundation                 0x1977c1328 __CFRunLoopRun + 968
53 CoreFoundation                 0x1977c0c34 CFRunLoopRunSpecific + 424
54 GraphicsServices               0x1a190a38c GSEventRunModal + 160
55 UIKitCore                      0x19b8f322c UIApplicationMain + 1932
56 MeviiBibleRead                 0x1024122a8 main + 23 (BibleAudioPanelControlItem.swift:23)
57 libdyld.dylib                  0x197648800 start + 4
```

## Steps for others to Reproduce
I cannot reproduce this error locally on any devices.

## Code Sample
That is my code:
```
extension RLMRealm {
    typealias CompleteBlock = () -> Void

    func commitTransaction(block: CompleteBlock) {
        beginWriteTransaction()
        block()
        do {
            try commitWriteTransaction()
        } catch let error {
            if self.inWriteTransaction {
                self.cancelWriteTransaction()
            }
            print(error.localizedDescription)
        }
    }

    static func insertOrUpdateItem(_ item: RLMObject) {
        let realm = RLMRealm.default()
        realm.commitTransaction {
            realm.addOrUpdate(item)
        }
    }

    static func deleteItem(_ item: RLMObject) {
        let realm = RLMRealm.default()
        realm.commitTransaction {
            realm.delete(item)
        }
    }
}
```

## Version of Realm and Tooling

Realm framework version: 
- Realm (4.1.1):
    - Realm/Headers (= 4.1.1)
  - Realm/Headers (4.1.1)

Realm Object Server version: None

Xcode version: Version 12.0.1 (12A7300)

iOS/OSX version: macOS Catalina 10.15.7 (19H2) 、iOS 12 / 13 / 14

Dependency manager + version: None

**Comment by ericjordanmossman** [comment]
Based on the stack trace I think you could be hitting the assertion failure that was fixed in [4.4.1](https://github.com/realm/realm-cocoa/releases/tag/v4.4.1). If you can share console output or upgrade to 4.4.1 then we could confirm.

**Comment by Apach3Q** [comment]
> Based on the stack trace I think you could be hitting the assertion failure that was fixed in [4.4.1](https://github.com/realm/realm-cocoa/releases/tag/v4.4.1). If you can share console output or upgrade to 4.4.1 then we could confirm.

Is there any problem with my code? The insert or delete method in extension.

**Comment by ericjordanmossman** [comment]
Depends what your goals are. If you're using swift you might be better off with our RealmSwift API directly in your view controllers instead of extending RLMRealm.
With respect to the crash, does 4.4.1 resolve the crash?

**Comment by Apach3Q** [comment]
> Depends what your goals are. If you're using swift you might be better off with our RealmSwift API directly in your view controllers instead of extending RLMRealm.
> With respect to the crash, does 4.4.1 resolve the crash?

Thank you for your advice. Because my project was transferred from OC and I will replace it with RealmSwift in future release.
Yesterday I released a new version to fix the problem by upgrade to 4.4.1.
Now I’m waiting for firebase crashlytics logs.
